### ISSUE TITLE
Tracking issue for future-incompatibility lint `deprecated_cfg_attr_crate_type_name`

### ISSUE THREAD

**Issue by bjorn3** [rust-lang/rust#91632]
### What is this lint about

The `deprecated_cfg_attr_crate_type_name` lint detects uses of the `#![cfg_attr(..., crate_type = "...")]` and `#![cfg_attr(..., crate_name = "...")]` attributes to conditionally specify the crate type and name in the source code. For example:

```rust
#![cfg_attr(debug_assertions, crate_type = "lib")]
```

### Explanation

The `#![crate_type]` and `#![crate_name]` attributes require a hack in the compiler to be able to change the used crate type and crate name after macros have been expanded. Neither attribute works in combination with Cargo as it explicitly passes `--crate-type` and `--crate-name` on the commandline. These values must match the value used in the source code to prevent an error.

### How to fix this warning/error

To fix the warning use `--crate-type` on the commandline when running rustc instead of `#![cfg_attr(..., crate_type = "...")]` and `--crate-name` instead of `#![cfg_attr(..., crate_name = "...")]`.

We are interested to hear about any/all use cases for conditional `#![crate_type]` and `#![crate_name]` attributes.

### Current status
* [x] https://github.com/rust-lang/rust/pull/83744 introduces the `deprecated_cfg_attr_crate_type_name` lint as warn-by-default
* [x] #99784 makes the `deprecated_cfg_attr_crate_type_name` lint deny-by-default
* [x] PR #129670 makes the `deprecated_cfg_attr_crate_type_name` lint a hard error

**Comment by est31** [comment]
I've filed a PR to make this deny-by-default: #99784

**Comment by est31** [comment]
Since the grep.app [seach conducted in the original PR](https://github.com/rust-lang/rust/pull/83744#issuecomment-825551806) there have been two new crates using it, despite the lint:

[wasmer](https://github.com/wasmerio/wasmer/commit/fcc9e741c3aee747b2c91575171a1c206127786b):
```Rust
#![cfg_attr(feature = "js", crate_type = "cdylib")]
#![cfg_attr(feature = "js", crate_type = "rlib")]
```
[azul](https://github.com/fschutt/azul/commit/2c1b2ed851ba5977e0f7636663884d6c5cec84c4#diff-2486c782348527d5e8a121166f399ed1449c5960b8079859b573cbc5f8b184c2R3):
```Rust
#![cfg_attr(feature ="cdylib", crate_type = "cdylib")]
#![cfg_attr(feature ="staticlib", crate_type = "staticlib")]
#![cfg_attr(feature ="rlib", crate_type = "rlib")]
```

Both uses were added in 2021, after #83744 was opened and before it was merged. Also both only modify `crate_type`. As @bjorn3 has asked above about use cases for `crate_type` within `cfg_attr()`, it might be interesting to think about alternatives one can suggest to the maintainers of these libraries.

Maybe one could extend cargo to support [`cargo:rustc-crate-type` in build.rs scripts](https://doc.rust-lang.org/cargo/reference/build-scripts.html)? Alternatively, one could maybe suggest adding a crate that has `crate-type` set in Cargo.toml, and then `include!()`s the `lib.rs`?

**Comment by bjorn3** [comment]
Wouldn't `cargo build --crate-type cdylib` suffice in both cases?

**Comment by est31** [comment]
It should actually, good point. Maybe the `wasm-pack` equivalent of it for `wasmer`, but yes. Also as you've wrote earlier, `#![crate_type = "..."]` is [totally ignored if `--crate-type` is present](https://github.com/rust-lang/rust/pull/83744#issuecomment-811855634) (only then though, if it's not present it's not ignored, but it seems that cargo always sets it?? not 100% sure). Ideally one would interact with downstream though before proposing a PR to remove.

**Comment by ia0** [comment]
> Wouldn't `cargo build --crate-type cdylib` suffice in both cases?

This doesn't work for me. My understanding is that `--crate-type` is only a `rustc` flag, you can't use it with `cargo build`. I would really like it to work with `cargo build` though. Is this planned?

**Comment by bjorn3** [comment]
Right, `--crate-type` is a flag for `cargo rustc`, not `cargo build`. My bad. `cargo rustc` should work in most cases that `cargo build` does. The main limitation is that it requires a single target within a single package rather than allowing multiple packages and targets.

**Comment by ia0** [comment]
Oh that's very interesting, I didn't know that. I always assumed `cargo rustc` was just a wrapper around a single invocation of `rustc` requiring you to compile dependencies separately and pass them on the command line. That's good to know, that should solve my issue. I think there's almost no cases where `cargo build` is useful to me then since I always use a single package (I never use workspaces) and always compile for a single target.

**Comment by est31** [comment]
Nice, seems the azul usage has been removed in the meantime: https://github.com/fschutt/azul/commit/6096a1b00e787764fe9bef21fdaa9699f24aba16#diff-2486c782348527d5e8a121166f399ed1449c5960b8079859b573cbc5f8b184c2L3

Now only the usage in wasmer is remaining. It has been reduced to a single attr:

https://github.com/wasmerio/wasmer/blob/ccbe870770604ecb61f2d8e0fa298504381f6784/lib/api/src/lib.rs#L29

I wonder whether it would be possible to enable `cdylib` on the api crate unconditionally?

**Comment by nyurik** [comment]
Note that the docs still [suggest this attribute](https://doc.rust-lang.org/reference/linkage.html)

**Comment by bjorn3** [comment]
Using plain `#![crate_name = "..."]` and `#![crate_type = "..."]` is fine. Using them behind a `#![cfg_attr(...)]` triggers this lint.

**Comment by nyurik** [comment]
Is there a recommended way for feature-controlled `crate-type`, e.g. with a `build.rs`?  A crate with optional `cdylib` type would also need to have a feature-gated panic handler and possibly other things, so it would still have to have a `cdylib` feature in Cargo.toml. So one would have to run this -- a probable source of bugs?

```shell
RUSTFLAGS='--crate-type cdylib' cargo build --features cdylib
```

P.S. Good point @bjorn3, thx.

**Comment by bjorn3** [comment]
`cargo rustc` accepts `--crate-type` to override the crate type. A build script can't change the crate type and `#![crate_type]` doesn't work as youbl did intend with cargo anyway. Rustc requires that `--crate-type` and `#![crate_type]` match up if both are specified afaik and cargo always passes `--crate-type`.

**Comment by bjorn3** [comment]
With `cargo rustc --crate-type cdylib --features cdylib` I think you don't need a separate mylib-cdylib crate.

**Comment by traviscross** [comment]
@rustbot labels +I-lang-nominated

Let's discuss whether to make this a hard error in Rust 2024.

**Comment by delta4chat** [comment]
The problem is that sometimes `dylib`, `cdylib`, or `staticlib` are not available (and cause compilation errors).
The only way to resolve this problem, that is, make rlib the default option, and to enable `dylib` under certain conditions, e.g. with `#[no_std]`:
```
error: "#[panic_handler]" function required, but not found

error: language item required, but not found: "eh_personality"
  |
  = note: this can occur when a binary crate with "#![no_std]" is compiled for a target where "eh_personality" is defined in the standard library
  = help: you may be able to compile for a target that doesn't need "eh_personality", specify a target with "--target" or in ".cargo/config"
```

So try doing something like this:
```
#![cfg_attr(feature="std", crate_type = "cdylib")]
```
Only attempt to compile `cdylib` library format if the std feature is available.

**Comment by delta4chat** [comment]
Or, can I just write my own `make` script and use `cargo rustc` to handle this situation?
Or can someone tell me a better alternative?

**Comment by ia0** [comment]
> Or, the only way I think, that is just write my own `make` script and use `cargo rustc` to handle this situation?

That's what was suggested in https://github.com/rust-lang/rust/issues/91632#issuecomment-1477975699 and what [I'm doing](https://github.com/google/wasefire/blob/5f6a744a15b0fae83a78953f5a12dfd37193026a/crates/xtask/src/main.rs#L365-L374) since then. And I didn't have any issue with it so far, so would definitely recommend if that's something your project would permit.

**Comment by delta4chat** [comment]
> and what [I'm doing](https://github.com/google/wasefire/blob/5f6a744a15b0fae83a78953f5a12dfd37193026a/crates/xtask/src/main.rs#L365-L374) 

Yes, this solution is actually similar to a make script, except that it's written in Rust. ~~Rust calls Rust to compile Rust hahaha.~~

So if we envision an elegant way to handle this situation without using this "make script" approach, it might be to add this logic to `Cargo.toml`, such as "Conditionally Set Rustc Compilation Flags Based On Features".

**Comment by est31** [comment]
I don't think there is any "elegant" solution to this problem, but there is a bunch of solutions:

* You can specify the crate type manually using `cargo rustc`. Usually you require specific crate types when integrating the project somewhere,

**Comment by delta4chat** [comment]
> I don't think there is any "elegant" solution

Yes, that's why it's important to use a bash script or Makefile to do this (because Cargo can't do it). It also means that Cargo can't cover all use cases.

> I think it's fine to live without an "elegant" solution because the use cases are pretty niche.

so `no-std & dylib` are pretty niche?

----

The issue is that even if you use `panic="abort"` in Cargo.toml, then Rustc will still requires libunwind and eh_personality: https://github.com/rust-lang/rust/issues/106864 / https://github.com/rust-lang/rust/issues/104159 / https://github.com/rust-lang/rust/issues/97602 / https://github.com/rust-lang/rust/issues/120655

also this causes most of Rust programs needs dynamic linking, even you specify https://github.com/rust-lang/rust/issues/118400

this also breaks `cross-rs/android`: https://github.com/cross-rs/cross/issues/1222

but `-Z build-std-features=panic_immediate_abort` is still unstable/nightly, can not use it for stable Rust. https://github.com/rust-lang/rust/issues/115022


Since "not all use cases require panic support", I think the Rust compiler should completely disable "eh_personality" by default if `#![no_std]` specified. or `opt-in` / `opt-out` (by flag)

**Comment by est31** [comment]
I've filed a PR now to make the lint a hard error: #129670

There has been enough time for users to move to alternatives discussed in this thread: It's been two years since my earlier PR to make the deprecation lint warn by default has been merged: #99784 .

**Comment by est31** [comment]
Closing as #129670 has been merged. It's on the path to being a hard error in the 1.83.0 release of Rust.

**Comment by est31** [comment]
I've tried to work on the followup refactor to #129670 to remove mutable state from `Session` but found that it's already been done in #114622. With that PR, what is written in the explanation section of this issue is obsolete: there is no mutable state in `Session` any more. I'm wondering now about further justifications for the change, not from a "should we halt this now" point of view, but from a "how to explain it" view.

### directory for the incr comp session dir

In #83744 @bjorn3 writes:

> [I]f we want to add incremental compilation to macro expansion or even parsing, we need the StableCrateId to be created together with the Session or even earlier as incremental compilation determines the incremental compilation session dir based on the StableCrateId.

### general ordering concerns

With this change it's easier to determine the crate name earlier in the compilation, which might be useful for structuring the state. All one needs is to peek into the provided source file, with no other state needed. Obtaining the crate name and crate type are very dependency-free operations now.

**Comment by bjorn3** [comment]
I think it would now be possible to revert https://github.com/rust-lang/rust/pull/114622 in a way that doesn't re-introduce the global mutable state in TyCtxt. This would also allow https://github.com/rust-lang/rust/blob/06d261daf62620e3449ba451898c7ad9ae750f41/compiler/rustc_incremental/src/persist/fs.rs#L611-L612 to avoid recomputing the crate name and type. And the same for https://github.com/rust-lang/rust/blob/06d261daf62620e3449ba451898c7ad9ae750f41/compiler/rustc_interface/src/util.rs#L466-L470 (this function doesn't handle `#![crate_type]` correctly at all right now.) And https://github.com/rust-lang/rust/blob/06d261daf62620e3449ba451898c7ad9ae750f41/compiler/rustc_driver_impl/src/lib.rs#L752-L753 and https://github.com/rust-lang/rust/blob/06d261daf62620e3449ba451898c7ad9ae750f41/compiler/rustc_driver_impl/src/lib.rs#L766 can also avoid recomputing the crate name.

**Comment by fmease** [comment]
> This would also allow https://github.com/rust-lang/rust/blob/06d261daf62620e3449ba451898c7ad9ae750f41/compiler/rustc_incremental/src/persist/fs.rs#L611-L612 to avoid recomputing the crate name and type.

Incidentally my open PR #127581 does away with recomputing the crate name inside `fs.rs`.

**Comment by est31** [comment]
@bjorn3 @fmease thank you!
